numpy/core/include/numpy/npy_common.h

```diff
@@ -361,49 +363,56 @@ typedef double npy_double;
 typedef Py_hash_t npy_hash_t;
 #define NPY_SIZEOF_HASH_T NPY_SIZEOF_INTP
 
-/*
- * Disabling C99 complex usage: a lot of C code in numpy/scipy rely on being
- * able to do .real/.imag. Will have to convert code first.
- */
-#if 0
-#if defined(NPY_USE_C99_COMPLEX) && defined(NPY_HAVE_COMPLEX_DOUBLE)
-typedef complex npy_cdouble;
-#else
-typedef struct { double real, imag; } npy_cdouble;
-#endif
-
-#if defined(NPY_USE_C99_COMPLEX) && defined(NPY_HAVE_COMPLEX_FLOAT)
-typedef complex float npy_cfloat;
-#else
-typedef struct { float real, imag; } npy_cfloat;
-#endif
-
-#if defined(NPY_USE_C99_COMPLEX) && defined(NPY_HAVE_COMPLEX_LONG_DOUBLE)
-typedef complex long double npy_clongdouble;
-#else
-typedef struct {npy_longdouble real, imag;} npy_clongdouble;
-#endif
-#endif
 #if NPY_SIZEOF_COMPLEX_DOUBLE != 2 * NPY_SIZEOF_DOUBLE
 #error npy_cdouble definition is not compatible with C99 complex definition ! \
         Please contact NumPy maintainers and give detailed information about your \
         compiler and platform
 #endif
-typedef struct { double real, imag; } npy_cdouble;
-
 #if NPY_SIZEOF_COMPLEX_FLOAT != 2 * NPY_SIZEOF_FLOAT
 #error npy_cfloat definition is not compatible with C99 complex definition ! \
         Please contact NumPy maintainers and give detailed information about your \
         compiler and platform
 #endif
-typedef struct { float real, imag; } npy_cfloat;
-
 #if NPY_SIZEOF_COMPLEX_LONGDOUBLE != 2 * NPY_SIZEOF_LONGDOUBLE
 #error npy_clongdouble definition is not compatible with C99 complex definition ! \
         Please contact NumPy maintainers and give detailed information about your \
         compiler and platform
 #endif
-typedef struct { npy_longdouble real, imag; } npy_clongdouble;
+
+
+#if defined(_MSC_VER) && !defined(__INTEL_COMPILER) && defined(__cplusplus)
+typedef struct 
+{
+    double _Val[2];
+} npy_cdouble;
+
+typedef struct
+{
+    float _Val[2];
+} npy_cfloat;
+
+typedef struct
+{
+    long double _Val[2];
+} npy_clongdouble;
+#elif defined(_MSC_VER) && !defined(__INTEL_COMPILER) /* && !defined(__cplusplus) */
+#include <complex.h>
+typedef _Dcomplex npy_cdouble;
+typedef _Fcomplex npy_cfloat;
+typedef _Lcomplex npy_clongdouble;
+#else /* !defined(_MSC_VER) || defined(__INTEL_COMPILER) */
+#ifdef __cplusplus
+extern "C++" {
+#endif
+#include <complex.h>
+#ifdef __cplusplus
+}
+#endif
+#undef complex
```

Comment:
I don't think this `#undef` should be here.  This breaks the use of `complex` in extensions modules that use `#include <complex.h>` that also include this NumPy header.

Reply:
xref #24343